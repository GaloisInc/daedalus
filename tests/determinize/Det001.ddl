-- Testing LL(1) with $[_] and some semantic value

def Main = { x = Simple1 ; y = SimpleNot1; z = Simple2; t = Simple3 }

def Simple1 =
    $['a']
  | $['b']
  | $['c']
  | $['d']

def SimpleNot1 =
    $['e']
  | $['f']
  | $['f']
  | $['g']

def Simple2 =
     @$['A']
  <| @$['B']
  <| @$['C']

def Simple3 =
     { @x = $['a'] ; @$['b']; ^ x }
  <| $['b']
  <| { @x = $['c'] ; @$['b']; ^ x + x }

