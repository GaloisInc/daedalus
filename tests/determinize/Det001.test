Det001.ddl
--dump-core
--determinize
