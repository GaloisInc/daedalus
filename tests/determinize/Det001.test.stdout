[WARNING] ./Det001.ddl:12:5--14:10:
  Using unbiased choice may be inefficient.
module DaedalusMain where
 
 
type Main =
  struct
    x : Word 8
    y : Word 8
    z : ()
    t : Word 8
 
-------------
 
-- ./Det001.ddl:5:1--9:10
proc Simple1() : Word 8 =
  -- ./Det001.ddl:6:5--9:10
  -- DETERMINIZE 1 Fully
  do _x0 <- match1 { ... }
     case _x0 of
       97 ->
         -- ./Det001.ddl:6:5--8:10
         -- ./Det001.ddl:6:5--7:10
         -- ./Det001.ddl:6:5--6:10
         pure (97 @(Word 8))
       98 ->
         -- ./Det001.ddl:6:5--8:10
         -- ./Det001.ddl:6:5--7:10
         -- ./Det001.ddl:7:5--7:10
         pure (98 @(Word 8))
       99 ->
         -- ./Det001.ddl:6:5--8:10
         -- ./Det001.ddl:8:5--8:10
         pure (99 @(Word 8))
       100 ->
         -- ./Det001.ddl:9:5--9:10
         pure (100 @(Word 8))
       _ ->
         fail_sys @(Word 8)
 
-- ./Det001.ddl:11:1--15:10
proc SimpleNot1() : Word 8 =
  -- ./Det001.ddl:12:5--15:10
  -- DETERMINIZE 1
  do _x0 <- match1 { ... }
     case _x0 of
       101 ->
         -- ./Det001.ddl:12:5--14:10
         -- ./Det001.ddl:12:5--13:10
         -- ./Det001.ddl:12:5--12:10
         pure (101 @(Word 8))
       102 ->
         -- OR RECONSTRUCTION
         -- ./Det001.ddl:12:5--14:10
           fork
             -- ./Det001.ddl:12:5--13:10
             -- ./Det001.ddl:13:5--13:10
             pure (102 @(Word 8))
           fork
             -- ./Det001.ddl:14:5--14:10
             pure (102 @(Word 8))
       103 ->
         -- ./Det001.ddl:15:5--15:10
         pure (103 @(Word 8))
       _ ->
         fail_sys @(Word 8)
 
-- ./Det001.ddl:17:1--20:12
proc Simple2() : () =
  -- ./Det001.ddl:18:6--20:12
  -- DETERMINIZE 1 Fully
  do _x0 <- match1 { ... }
     case _x0 of
       65 ->
         -- ./Det001.ddl:18:6--19:12
         -- ./Det001.ddl:18:7--18:12
         pure ()
       66 ->
         -- ./Det001.ddl:18:6--19:12
         -- ./Det001.ddl:19:7--19:12
         pure ()
       67 ->
         -- ./Det001.ddl:20:7--20:12
         pure ()
       _ ->
         fail_sys @()
 
-- ./Det001.ddl:22:1--25:39
proc Simple3() : Word 8 =
  -- ./Det001.ddl:23:6--25:39
  -- DETERMINIZE 1 Fully
  do _x0 <- match1 { ... }
     case _x0 of
       97 ->
         -- ./Det001.ddl:23:6--24:11
         -- ./Det001.ddl:23:9--23:33
         -- ./Det001.ddl:23:13--23:18
         do let x = 97 @(Word 8)
            -- ./Det001.ddl:23:22--23:33
            -- ./Det001.ddl:23:23--23:28
            match1_ { 98 @(Word 8) }
            -- ./Det001.ddl:23:31--23:33
            -- ./Det001.ddl:23:31--23:33
            pure x
       98 ->
         -- ./Det001.ddl:23:6--24:11
         -- ./Det001.ddl:24:6--24:11
         pure (98 @(Word 8))
       99 ->
         -- ./Det001.ddl:25:9--25:37
         -- ./Det001.ddl:25:13--25:18
         do let x = 99 @(Word 8)
            -- ./Det001.ddl:25:22--25:37
            -- ./Det001.ddl:25:23--25:28
            match1_ { 98 @(Word 8) }
            -- ./Det001.ddl:25:31--25:37
            -- ./Det001.ddl:25:31--25:37
            pure (x + x)
       _ ->
         fail_sys @(Word 8)
 
-- ./Det001.ddl:3:1--3:69
proc Main() : Main =
  -- ./Det001.ddl:3:14--3:69
  do x <- -- ./Det001.ddl:3:18--3:24
          Simple1()
     -- ./Det001.ddl:3:28--3:69
     y <- -- ./Det001.ddl:3:32--3:41
          SimpleNot1()
     -- ./Det001.ddl:3:44--3:69
     z <- -- ./Det001.ddl:3:48--3:54
          Simple2()
     -- ./Det001.ddl:3:57--3:69
     t <- -- ./Det001.ddl:3:61--3:67
          Simple3()
     -- ./Det001.ddl:3:12--3:69
     pure Main {x = x, y = y, z = z, t = t}
